pkg/policy/api/v3/rule.go

```diff
@@ -0,0 +1,240 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v3
+
+import (
+	"github.com/cilium/cilium/pkg/labels"
+)
+
+// Rule is a policy rule which must be applied to all endpoints which match the
+// labels contained in the endpointSelector
+//
+// Each rule is split into an ingress section which contains all rules
+// applicable at ingress, and an egress section applicable at egress. For rule
+// types such as `L4Rule` and `CIDR` which can be applied at both ingress and
+// egress, both ingress and egress side have to either specifically allow the
+// connection or one side has to be omitted.
+//
+// Either ingress, egress, or both can be provided. If both ingress and egress
+// are omitted, the rule has no effect.
+type Rule struct {
+	// IdentitySelector selects all endpoints which should be subject to
+	// this rule. Cannot be empty.
+	EndpointSelector IdentitySelector `json:"endpointSelector"`
+
+	// Ingress is a list of IngressRule which are enforced at ingress.
+	// If omitted or empty, this rule does not apply at ingress.
+	//
+	// +optional
+	Ingress []IngressRule `json:"ingress,omitempty"`
+
+	// Egress is a list of EgressRule which are enforced at egress.
+	// If omitted or empty, this rule does not apply at egress.
+	//
+	// +optional
+	Egress []EgressRule `json:"egress,omitempty"`
+
+	// Labels is a list of optional strings which can be used to
+	// re-identify the rule or to store metadata. It is possible to lookup
+	// or delete strings based on labels. Labels are not required to be
+	// unique, multiple rules can have overlapping or identical labels.
+	//
+	// +optional
+	Labels labels.LabelArray `json:"labels,omitempty"`
+
+	// Description is a free form string, it can be used by the creator of
+	// the rule to store human readable explanation of the purpose of this
+	// rule. Rules cannot be identified by comment.
+	//
+	// +optional
+	Description string `json:"description,omitempty"`
+}
+
+type IngressRule struct {
+	FromIdentities *IdentityRule `json:"fromIdentities,omitempty"`
+
+	FromRequires *EndpointRequire `json:"fromRequires,omitempty"`
+
+	FromCIDRs *CIDRRule `json:"fromCIDR,omitempty"`
+
+	FromEntities *EntityRule `json:"fromEntities,omitempty"`
+}
+
+type EgressRule struct {
+	ToIdentities *IdentityRule `json:"toIdentities,omitempty"`
+
+	ToRequires *EndpointRequire `json:"toRequires,omitempty"`
+
+	ToCIDRs *CIDRRule `json:"toCIDR,omitempty"`
+
+	ToEntities *EntityRule `json:"toEntities,omitempty"`
+
+	ToServices *ServiceRule `json:"toServices,omitempty"`
```

Comment:
This is no longer a slice, should this be renamed ToService? It can still match multiple services via selector, so I am not sure about it.